lib/memo_wise.rb

```diff
@@ -163,25 +252,46 @@ def self.has_block_arg?(method) # rubocop:disable Naming/PredicateName
   def self.prepended(target) # rubocop:disable Metrics/PerceivedComplexity
     class << target
       # NOTE: See YARD docs for {.memo_wise} directly below this method!
-      def memo_wise(method_name) # rubocop:disable Metrics/PerceivedComplexity
+      def memo_wise(method_name_or_hash) # rubocop:disable Metrics/PerceivedComplexity
+        case method_name_or_hash
+        when Symbol
+          method_name = method_name_or_hash
+        when Hash
+          unless method_name_or_hash.keys == [:self]
+            raise ArgumentError,
+                  "`:self` is the only key allowed in memo_wise"
+          end
+
+          method_name = method_name_or_hash[:self]
+        end
+
+        # In Ruby, "class methods" are implemented as normal instance methods
+        # on the "singleton class" of a given Class object, found via
+        # {Class#singleton_class}.
+        # See: https://medium.com/@leo_hetsch/demystifying-singleton-classes-in-ruby-caf3fa4c9d91
+        klass = self
+        singleton = klass.singleton_class
+
+        if singleton.private_instance_methods.include?(method_name) ||
```

Comment:
On the previous PR, you (@jemmaissroff) had commented that we might be able to just use

```ruby
if singleton.instance_method(method_name)
```

Reply:
So if it's not an instance method, it'll raise a `NameError`, which means the path I see is to surround this in a `begin / rescue` block. I think that's a little less clean than checking if it's in `private_instance_methods` or `instance_methods`. Also, in digging into this, I realized we didn't have test coverage of private class methods, so I added that

Reply:
I think we need test coverage that we can have the same `method_name` at both class and instance levels, and that here we should be using the presence of the `self:` keyword arg to choose which one to memoize?